serviceregistration/kubernetes/client/client.go

```diff
@@ -0,0 +1,285 @@
+	}
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			if c.logger.IsWarn() {
+				// Failing to close response bodies can present as a memory leak so it's
+				// important to surface it.
+				c.logger.Warn(fmt.Sprintf("unable to close response body: %s", err))
+			}
+		}
+	}()
+
+	// If we're not supposed to read out the body, we have nothing further
+	// to do here.
+	if ptrToReturnObj == nil {
+		return nil
+	}
+
+	// Attempt to read out the body into the given return object.
+	return json.NewDecoder(resp.Body).Decode(ptrToReturnObj)
+}
+
+func (c *Client) getCheckRetry(req *http.Request) retryablehttp.CheckRetry {
+	return func(ctx context.Context, resp *http.Response, err error) (bool, error) {
+		switch resp.StatusCode {
+		case 200, 201, 202, 204:
+			// Success.
+			return false, nil
+		case 401, 403:
+			// Perhaps the token from our bearer token file has been refreshed.
+			config, err := inClusterConfig()
+			if err != nil {
+				return false, err
+			}
+			if config.BearerToken == c.config.BearerToken {
+				// It's the same token.
+				return false, fmt.Errorf("bad status code: %s", sanitizedDebuggingInfo(req, resp.StatusCode))
+			}
+			c.config = config
+			// Continue to try again, but return the error too in case the caller would rather read it out.
+			return true, fmt.Errorf("bad status code: %s", sanitizedDebuggingInfo(req, resp.StatusCode))
+		case 404:
+			return false, &ErrNotFound{debuggingInfo: sanitizedDebuggingInfo(req, resp.StatusCode)}
+		case 500, 502, 503, 504:
+			// Could be transient.
+			return true, fmt.Errorf("unexpected status code: %s", sanitizedDebuggingInfo(req, resp.StatusCode))
+		}
+		// Unexpected.
+		return false, fmt.Errorf("unexpected status code: %s", sanitizedDebuggingInfo(req, resp.StatusCode))
+	}
+}
+
+type Pod struct {
+	Metadata *Metadata `json:"metadata,omitempty"`
+}
+
+type Metadata struct {
+	Name string `json:"name,omitempty"`
+
+	// This map will be nil if no "labels" key was provided.
+	// It will be populated but have a length of zero if the
+	// key was provided, but no values.
+	Labels map[string]string `json:"labels,omitempty"`
+}
+
+type PatchOperation string
+
+const (
+	Unset   PatchOperation = "unset"
+	Add                    = "add"
+	Replace                = "replace"
+)
+
+type Patch struct {
+	Operation PatchOperation
+	Path      string
+	Value     interface{}
+}
+
+type ErrNotFound struct {
+	debuggingInfo string
+}
+
+func (e *ErrNotFound) Error() string {
+	return e.debuggingInfo
+}
+
+// sanitizedDebuggingInfo provides a returnable string that can be used for debugging. This is intentionally somewhat vague
+// because we don't want to leak secrets that may be in a request or response body.
+func sanitizedDebuggingInfo(req *http.Request, respStatus int) string {
+	return fmt.Sprintf("req method: %s, req url: %s, resp statuscode: %d", req.Method, req.URL, respStatus)
+}
+
+// stopChContext allows us to cause requests to stop retrying if we receive
```

Comment:
In lieu of implementing the whole context interface from scratch, could you do something like this in your `do` function:

```
...
	ctx, cancelFunc := context.WithCancel(context.Background())
	go func(ch <-chan struct{}) {
		<-ch
		cancelFunc()
	}(c.stopCh)

	retryableReq.WithContext(ctx)
...
```

Reply:
Great idea! I knew there had to be a simpler way rather than re-implementing the exhaustive Context interface.

Reply:
This is done here: https://github.com/hashicorp/vault/pull/8249/commits/630f180f87b9746129605e9ae7ca10628d5d17e5